great_expectations/rule_based_profiler/estimators/bootstrapped_standard_error_optimization_based_estimator.py

```diff
@@ -0,0 +1,356 @@
+        # This quantile is used in the calculation of the number of bootstrap samples as a function of the standard
+        # variance of the deviation (which, in turn, depends on the excess kurtosis of the distribution underlying the
+        # bootstrapped samples of the statistic.  The fact that one half of this probability factors in the equation
+        # points to the symmetric nature of the Normal distribution.  (For background on the quantile function of the
+        # standard Normal distribution, ses https://en.wikipedia.org/wiki/Normal_distribution and references therein.)
+        self._quantile_complement_prob_outside_bound_divided_by_2: np.float64 = (
+            scipy.stats.norm.ppf(
+                1.0 - self._prob_bootstrapped_statistic_deviation_outside_bound / 2.0
+            )
+        )
+
+        self._optimal_num_bootstrap_samples_estimations = []
+
+    def compute_bootstrapped_statistic_samples(self) -> np.ndarray:
+        optimal_num_bootstrap_samples: int = (
+            self._estimate_optimal_num_bootstrap_samples()
+        )
+        bootstrap_samples: np.ndarray = self._generate_bootstrap_samples(
+            num_bootstrap_samples=optimal_num_bootstrap_samples
+        )
+        return bootstrap_samples
+
+    def _estimate_optimal_num_bootstrap_samples(
+        self,
+    ) -> int:
+        # There are no bootstrapped samples initially; hence, assume zero excess kurtosis (i.e., Normal distribution).
+        optimal_num_bootstrap_samples: int = self._estimate_min_num_bootstrap_samples(
+            bootstrap_samples=None
+        )
+        self._optimal_num_bootstrap_samples_estimations.append(
+            optimal_num_bootstrap_samples
+        )
+
+        previous_max_optimal_num_bootstrap_samples: int = 0
+        current_max_optimal_num_bootstrap_samples: int = max(
+            self._optimal_num_bootstrap_samples_estimations
+        )
+
+        idx: int = 2
+        # Iterate until the maximum of the list of minimum estimates of the number of bootstrapped samples converges.
+        while (
+            current_max_optimal_num_bootstrap_samples
+            > previous_max_optimal_num_bootstrap_samples
+        ):
+            bootstrap_samples = self._generate_bootstrap_samples(
+                num_bootstrap_samples=optimal_num_bootstrap_samples
+            )
+            optimal_num_bootstrap_samples = self._estimate_min_num_bootstrap_samples(
+                bootstrap_samples=bootstrap_samples
+            )
+            self._optimal_num_bootstrap_samples_estimations.append(
+                optimal_num_bootstrap_samples
+            )
+            previous_max_optimal_num_bootstrap_samples = (
+                current_max_optimal_num_bootstrap_samples
+            )
+            current_max_optimal_num_bootstrap_samples = max(
+                self._optimal_num_bootstrap_samples_estimations
+            )
+            idx = idx + 1
```

Comment:
Do you think it makes sense @alexsherstinsky to limit the number of iterations here? Perhaps in a user-configurable way? (I know it's limited by the while loop condition, but do we need to protect from non-convergence / infinite loop?)

Reply:
@anthonyburdi This is a very good question.  I am going to add a "just in case" limit in the upcoming commit (and will let you know), but this should not be necessary.  The algorithm is guaranteed to converge -- and in my experiments has done so in under 5 steps (2 of which are present a priori).  This is because the kurtosis of the distribution is determined after only a few tries, and the number of bootstrap samples is proportional to the maximum kurtosis.  The maximum stops changing very rapidly.  Thanks!

Reply:
Thank you for clarifying! That makes sense.

Reply:
🙇